pkg/deviceplugin/plugin.go

```diff
@@ -0,0 +1,133 @@
+package deviceplugin
+
+import (
+	"fmt"
+
+	"github.com/golang/glog"
+	"golang.org/x/net/context"
+	pluginapi "k8s.io/kubernetes/pkg/kubelet/apis/deviceplugin/v1beta1"
+
+	"github.com/kubevirt/macvtap-cni/pkg/util"
+)
+
+const (
+	tapPath         = "/dev/tap"
+	suffix          = "Mvp"
+	defaultCapacity = 100
+)
+
+type MacvtapDevicePlugin struct {
+	Name        string
+	Master      string
+	Mode        string
+	Capacity    int
+	stopWatcher chan struct{}
+}
+
+func NewMacvtapDevicePlugin(name string, master string, mode string, capacity int) *MacvtapDevicePlugin {
+	return &MacvtapDevicePlugin{
+		Name:        name,
+		Master:      master,
+		Mode:        mode,
+		Capacity:    capacity,
+		stopWatcher: make(chan struct{}),
+	}
+}
+
+func (mdp *MacvtapDevicePlugin) generateMacvtapDevices() []*pluginapi.Device {
+	var macvtapDevs []*pluginapi.Device
+
+	var capacity = mdp.Capacity
+	if capacity <= 0 {
+		capacity = defaultCapacity
+	}
+
+	for i := 0; i < capacity; i++ {
+		name := fmt.Sprint(mdp.Name, suffix, i)
+		macvtapDevs = append(macvtapDevs, &pluginapi.Device{
+			ID:     name,
+			Health: pluginapi.Healthy,
+		})
+	}
+	return macvtapDevs
+}
+
+func masterExists(master string) bool {
+	if err := util.LinkExists(master); err != nil {
+		glog.Warningf("Master %s not found: %v", master, err)
+		return false
+	}
+	return true
+}
+
+func (mdp *MacvtapDevicePlugin) ListAndWatch(e *pluginapi.Empty, s pluginapi.DevicePlugin_ListAndWatchServer) error {
```

Comment:
Please add a comment describing which interfaces are offered here. I don't get why it requires so much code, you just need to create bunch of fake Devices, no?

Reply:
Not only that but also keeping up with the master device. If there is no master, there is no point in offering devices for allocation as it will fail.